contracts/logion_psp34/lib.rs

```diff
@@ -0,0 +1,111 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+#[cfg(test)]
+pub mod tests;
+
+#[openbrush::implementation(PSP34, PSP34Mintable, PSP34Metadata)]
+#[openbrush::contract]
+pub mod logion_psp34 {
+    use core::fmt::{Display, Error, Formatter};
+    use core::str::FromStr;
+    use ink_env::format;
+    use ink_env::hash::Sha2x256;
+    use openbrush::traits::Storage;
+
+    const ID_FOR_LOGION_METADATA: &Id = &Id::U8(1);
+    const NONCE_KEY: &'static str = "nonce";
+    const COLLECTION_LOC_ID_KEY: &'static str = "collection_loc_id";
+    const CERT_HOST_KEY: &'static str = "cert_host";
+
+    pub struct LogionHash {
+        value: Hash
+    }
+
+    impl Display for LogionHash {
+
+        fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
+            write!(f, "0x")?;
+            for i in self.value.as_ref() {
+                write!(f, "{:02x?}", i)?;
+            }
+            Ok(())
+        }
+    }
+
+    #[ink(storage)]
+    #[derive(Default, Storage)]
+    pub struct LogionPsp34 {
+        #[storage_field]
+        psp34: psp34::Data,
+        #[storage_field]
+        metadata: metadata::Data,
+    }
+
+    impl LogionPsp34 {
+        #[ink(constructor)]
+        pub fn new(nonce: String, collection_loc_id: u128, cert_host: String) -> Self {
+            let mut instance = Self::default();
+
+            metadata::Internal::_set_attribute(&mut instance, ID_FOR_LOGION_METADATA.clone(), String::from(NONCE_KEY), nonce);
+            metadata::Internal::_set_attribute(&mut instance, ID_FOR_LOGION_METADATA.clone(), String::from(COLLECTION_LOC_ID_KEY), format!("{}", collection_loc_id));
+            metadata::Internal::_set_attribute(&mut instance, ID_FOR_LOGION_METADATA.clone(), String::from(CERT_HOST_KEY), cert_host);
```

Comment:
The Logion parameters (none, LOC ID and host) are the same for all tokens so setting them for each one is quite sub-optimal. What about adding a storage "section" for them (something like `logion: logion::Data` in the storage struct)?

Reply:
As discussed, will be part of a next PR.